### ISSUE TITLE
Not Syncs

### ISSUE THREAD

**Issue by handicraftsman** [micheleg/dash-to-dock#293]
Dock not syncs running applications list (which not in favorites). For example: i am starting Powder Toy (google, if don't know), but it shows on dock only when i am disable and reenable option "Show running applications" (or after some time). Same thing with closing applications.

**Comment by handicraftsman** [comment]
Ohhh, new info: shell restart fixes this for a while...

**Comment by micheleg** [comment]
Is it specific to this or some applications? Also, does the problem occurs with the extension disabled in the normal dock inside the overview?

**Comment by micheleg** [comment]
I forgot to ask: are you running wayland?

**Comment by handicraftsman** [comment]
Nope, running X. And not tested disabling extension

**Comment by poxin13** [comment]
I'm having this same issue with v53 on Ubuntu 16.04/Gnome. Disabling/Enabling show running applications syncs them but you have to do that manually - odd.

**Comment by micheleg** [comment]
Sorry if I'm tedious about this: Does the problem occurs with the extension disabled, i.e. in the normal dock inside the overview?

**Comment by poxin13** [comment]
Well for me, this was fixed on a reboot and seems to be working fine now. Can't speak for @handicraftsman. I made some tweaks before hand but not exactly sure what step may have impacted it.

**Comment by sbtypo3** [comment]
I have this problem, steam showed up 4 or 5 minutes after the app started and I was able to use it. I have seen this with atom editor too.

Ubutu gnome 16.04 dash to dock in panel mode

**Comment by franglais125** [comment]
Maybe adding a signal to `_queueRedisplay()` might help with this. I remember seeing this issue sporadically, but it went away once I started to use workspace isolation (which connects `"restacked"`).

@sbtypo3 if you see this issue again, can you try just removing* a favorite app? That should trigger a `_queueRedisplay()`. To be clear, I am not suggesting this as a workaround every time you have this problem, just to know if adding a signal would be helpful at all. If you see that the dash is refreshed properly, then perhaps the `_queueRedisplay()` thing could be the solution.

*By that I simply mean right clik on a favorite app and select "Remove from Favorites"

I'll try to reproduce the issue (without having workspace isolation activated), and try to see if this works.

**Comment by sbtypo3** [comment]
it happened again just now. I unfaved another app and steam showed up straight away

**Comment by franglais125** [comment]
@sbtypo3 Thanks for the feedback!
What version of the (1) extension and (2) Gnome Shell are you using?

**Comment by sbtypo3** [comment]
3.18.5 (Ubuntu Gnome 16.04) and gnome3 ppa to get latest version of nautilus, but not the staging ppa.

Dash-to-Dock 53

**Comment by franglais125** [comment]
@micheleg what do you think of this patch [fbd6079a3fc227c8344dfaeaa23b8a39de4ead17] ?

**Comment by micheleg** [comment]
You are just moving the signal from one class to the oher, or am I missing something?

**Comment by franglais125** [comment]
Yes, but in the _docking.js_ case, it is only connected with workspace isolation.
I suspect this is why the problem went away for me (I used to have back then). Moving to _dash.js_ activates it by default.

**Comment by franglais125** [comment]
@sbungert there is a new version available. Do you mind updating and checking if the problem is still present?

https://extensions.gnome.org/extension/307/dash-to-dock/

**Comment by sbungert** [comment]
It's sort of half fixed... now when the window for steam appears saying that it is loading and connecting as my user name it appears straight away on the dock, but when that closes and the main steam window opens there is no icon until i open another app or unfav something

**Comment by franglais125** [comment]
Thanks.

Here is a branch (with a single commit ahead of v54) with what I hope is a fix: https://github.com/franglais125/dash-to-dock/tree/bugfix_sync

Do you know how to test it?

**Comment by sbungert** [comment]
No... can you give me some tips?

**Comment by franglais125** [comment]
Alternatively, you can activate the "Isolate workspaces" option from the Dash-to-dock settings.

Not permanently, but just to see if the problem goes away, i.e. try opening Steam and see if the dock syncs properly when this option is activated.

**Comment by micheleg** [comment]
Again, can somebody that can reproduce the bug check if it occurs also with the default dock, .i.e. with this extension disabled? I have the feeling it is a problem with the shell. @franglais125's workaround look okish, but I'd rather know why it is required, and if it's the case file an upstream bug,

**Comment by sbungert** [comment]
I just turned it off in the tweak tool and tried steam. The loading window shows up automatically on the dash on overview but the main steam window that appears after signing into steam doesn't

**Comment by sbungert** [comment]
On ubuntugnome 16.10 this problem seems to have been fixed

**Comment by micheleg** [comment]
@franglais125 The problem with the proposed patch is that the restacked signals is emitted a bit too often to just use it as a workaround,  for instance anytime you change window.

**Comment by franglais125** [comment]
@micheleg Yes, I generally agree with that. As the other bug you just fixed (issue #386) showed, redisplay() is called way too often. At least for the workspace isolation to work well, the `restack` signal has worked well, although it does seem a bit excessive at times. I'll try and keep an eye on this and see if I can send improvements from time to time.

As for this issue, it seems that newer versions of the Shell do work better, so maybe over time this particular problem will slowly go away.

**Comment by micheleg** [comment]
To me it looks like an upstream bug in the appsystem, which maybe has been solved in recent versions.

**Comment by Gauravano** [comment]
I am using Ubuntu 17.10 and facing the same issue, the dock was working good in 17.04, but now doesn't show running applications.
